### DISCUSSION TITLE
Error handling and invalidating when combining providers

### DISCUSSION THREAD

**Discussion by markusrut** [rrousselGit/riverpod#2836] [Q&A]
Im having some issues in how to handle errors and invalidating provider states in cases where im combining provider states

To explain the issues im using one of the examples from the docs
https://docs-v2.riverpod.dev/docs/concepts/combining_providers

Having the combined providers as follows
```Dart
@riverpod
Future<List<Character>> characters(CharactersRef ref) async {
  final search = ref.watch(searchProvider);
  final configs = await ref.watch(configsProvider.future);
  final response = await dio.get<List<Map<String, dynamic>>>(
      '${configs.host}/characters?search=$search');

  return response.data!.map(Character.fromJson).toList();
}
```

With the following example widget for using the provider
```Dart

class TestWidget extends ConsumerWidget {
  const TestWidget({super.key});

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final asyncValue = ref.watch(charactersProvider);
    return asyncValue.when(
      data: (data) => Container(),
      error: (error, stackTrace) {
        return ErrorCard(
          error: error,
          stackTrace: stackTrace,
          onRetry: () => ref.invalidate(charactersProvider),
        );
      },
      loading: () => const CircularProgressIndicator(),
    );
  }
}
```
Whats important here is the ErrorCard widget that contains a retry button with the goal of invalidating the provider in order to try and refetch value in case of the error being temporary connection issue etc.

If in this example the issue origins from the `configsProvider` simply invalidating the `charactersProvider` that combines the states will not do anything since the dependencies in turn is not invalidated.

Is there any way that i can solve this issue without having the consuming widget being aware of the separate parts of the combinded providers.
I did not manage to solve this by having the `charactersProviders` implement an `onDispose` to in turn invalidate dependencies since this would retrigger when dependencies are changes because of the watch.

Has anyone done anything similar to this that could give me any hints?

Also facing the same issue when implementing `RefreshIndicator` to get pull to refresh functionality where only the combining the provider would get refreshed if the consumer is not aware of the inner dependencies

**Comment by AhmedLSayed9** [comment]
A good approach is to convert charactersProvider to class-based provider (AsyncNotifier) and add a method that invalidate the needed deps, then you'd just call `ref.read(charactersProvider.notifier).invalidate();`

Another approach is to let all deps watch some dummy provider (i.e: charactersDepsProvider) and just invalidate that provider. but that sounds weird imo and you might forget to watch the dummy provider when you add new dependency, while the first approach can be easily noticed since it is in the same class.

**Reply by markusrut** [reply]
Yeah, from what i can think of making it a class based and adding specific function to invalidate dependencies might be the best solution currently.
That does however require that the consumer of the provider knows that when using one with this special invalidation.

Feels like a more optimized solution would be to find a way to support this in onDisposed, by somehow knowing if the dispose is initiated by a dependency changing or a specific invalidate for the specific provider.
However adding a parameter to the onDisposed callback with this information would cause it to be a signature change that im assuming might not be wanted.

Another solution could be to add an optional parameter to ref.onDispose instead of adding it to the callback.
But without knowing how onDisposed is handled internally this might include other issues

**Reply by AhmedLSayed9** [reply]
I don't think accessing ref inside `onDispose` is a good idea.
I believe invalidating providers inside `onDispose` will be done in the next frame, if not throwing an error. It's the same situation of https://github.com/rrousselGit/riverpod/issues/2538

**Reply by markusrut** [reply]
The linked issue is from my understanding about the widgets dispose, my thinking is to have it in the providers dispose, but maby the issues are the same

**Reply by AhmedLSayed9** [reply]
Yep. I think it will face the same issues.

but, I agree that they should be a way to override the behavior of `invalidate` for a provider to avoid the need to remember calling some specific method which is error-prone.
